src/cargo/core/compiler/context/unit_dependencies.rs

```diff
@@ -38,27 +39,33 @@ fn deps_of<'a, 'b, 'cfg>(
     unit: &Unit<'a>,
     cx: &Context<'a, 'cfg>,
     deps: &'b mut HashMap<Unit<'a>, Vec<Unit<'a>>>,
+    profile_for: ProfileFor,
 ) -> CargoResult<&'b [Unit<'a>]> {
     if !deps.contains_key(unit) {
-        let unit_deps = compute_deps(unit, cx, deps)?;
-        deps.insert(*unit, unit_deps.clone());
-        for unit in unit_deps {
-            deps_of(&unit, cx, deps)?;
+        let unit_deps = compute_deps(unit, cx, deps, profile_for)?;
+        let to_insert: Vec<_> = unit_deps.iter().map(|&(unit, _)| unit).collect();
+        deps.insert(*unit, to_insert);
+        for (unit, profile_for) in unit_deps {
```

Comment:
If I add `assert_eq!(unit.profile_for, profile_for);` it is not triggered, so, perhaps, we don't need to return pairs after all?